### PR TITLE
PUBDEV-4639: Added requirement for data.table.

### PR DESCRIPTION
**Pull request by wendycwong**
This PR enable data.table by default as specified by JIRA: https://0xdata.atlassian.net/browse/PUBDEV-4639.

### CODE REVIEW

**Comment by wendycwong** [comment]
@honzasterba 

Thanks for the review.  I don't understand what that means.  Please help?  W

**Comment by wendycwong** [comment]
Honza:  I have added the change in h2o-3-DESCRIPTION.template.  Let me know if that is the right place.  If not, please point me to the right file.

**Comment by honzasterba** [comment]
@wendycwong no, I think we want to keep data.table in suggest since its not required, the data.table support should only enable itself when data.table is installed

**Comment by wendycwong** [comment]
@honzasterba Can you tell me exactly what I need to change then.  Thanks.  W

**Comment by honzasterba** [comment]
I change to DESCRIPTION should be reverted as I see you did.
Also I added a commit to this PR but it is gone so I added the change again.

**Comment by wendycwong** [comment]
Honza:  Thanks for the help.  Do you think this PR is ready to go?  Or more changes are needed?  I have been testing it continuously and it seems to work fine.  Let me know.  Thanks, Wendy

**Comment by honzasterba** [comment]
maybe check if there is some docs that need updating @angela0xdata

**Comment by wendycwong** [comment]
Honza:  There are two runits that are failing in RSmoke test.  That needs to be resolved before this PR can merge.

**Comment by michalkurka** [comment]
Needs fixing before merging:

```
[2020-02-13T01:04:38.981Z] * checking R code for possible problems

[2020-02-13T01:04:38.981Z] use.package: no visible global function definition for

[2020-02-13T01:04:38.981Z]   ‘installed.packages’

[2020-02-13T01:04:38.981Z] Undefined global functions or variables:

[2020-02-13T01:04:38.981Z]   installed.packages

[2020-02-13T01:04:38.981Z] Consider adding

[2020-02-13T01:04:38.981Z]   importFrom("utils", "installed.packages")

```

AFAIK @ledell offered to help with that

**Comment by honzasterba** [comment]
I have pushed a fix for the CRAN issue and also the actual change of the default value for data.table use. 
@mattdowle could you please take a look at the test failures?

**Comment by mattdowle** [comment]
@honzasterba The PR I submitted https://github.com/h2oai/h2o-3/pull/4308 does not appear to have been merged yet @wendycwong @ledell ?

**Comment by honzasterba** [comment]
@mattdowle I believe wendy cherry picked your commit into this PR, I am now working on fixing the rest of the tests

**Comment by mattdowle** [comment]
@honzasterba Ah ok I see the cherry pick now.  I was looking at the error in example(as.h2o) and I see your recent commit.   I see fail just now again : 
```
Error in .subset(x, j) : invalid subscript type 'list'
[2020-02-21T09:50:30.146Z] Calls: as.h2o ... .h2o.calc_stm_svm -> [ -> [.data.table -> [.data.frame
[2020-02-21T09:50:30.146Z] Execution halted
```
Did you write earlier above in this PR that data.table is downgraded to suggests in DESCRIPTION? If so that will affect using data.table syntax.  It seems that as.h2o.Matrix is using data.table syntax.  For that to work data.table needs to know h2o is data.table-aware.

**Comment by honzasterba** [comment]
yes, I have fixed the data-table aware issue yesterday, but it seems that @wendycwong did another rebase afterwards and my commit is lost :(

**Comment by wendycwong** [comment]
Sorry guys.  I will take my hands off this PR now.

**Comment by mattdowle** [comment]
This branch passes `R CMD check` locally for me. When running on Jenkins it seems to be running `example(as.h2o)` even though that code is inside a `\dontrun[ ... }` section.  Has anyone seen that before?

**Comment by honzasterba** [comment]
@mattdowle the changes that you reverted were my fixes for failing tests, we need those back

**Comment by mattdowle** [comment]
Failed due to Jenkins: `No space left on device`

**Comment by mattdowle** [comment]
@honzasterba  R CMD check passes for me, and this change fixes the failing example. Maybe you can take it from here then.

**Comment by honzasterba** [comment]
thanks @mattdowle , I will take it from here

File: h2o-core/src/main/java/water/api/DatasetServlet.java [hunk 1]
```diff
@@ -20,18 +20,17 @@
...
-      if (hex_string != null && hex_string.toLowerCase().equals("true")) {
-        use_hex = true;
-      }
-
       Frame dataset = DKV.getGet(f_name);
-      InputStream is = dataset.toCSV(new Frame.CSVStreamParams().setHexString(use_hex));
+      Frame.CSVStreamParams parms = new Frame.CSVStreamParams()
+          .setHexString(Boolean.parseBoolean(hex_string))
+          .setEscapeQuotes(Boolean.parseBoolean(escape_quotes_string));
+      InputStream is = dataset.toCSV(parms);
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/api/DatasetServlet.java]: Looks like the `is; is not closed anywhere, not even in `copyStream`. 

I offer the following patch that simplifies the following block by using try-with-resources:

```patch
Index: h2o-core/src/main/java/water/api/DatasetServlet.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- h2o-core/src/main/java/water/api/DatasetServlet.java	(revision 6bf5e494de07c7199d46a6d1c182ae2669af02ef)
+++ h2o-core/src/main/java/water/api/DatasetServlet.java	(date 1582538708720)
@@ -4,7 +4,6 @@
 import water.fvec.Frame;
 import water.server.ServletUtils;
 import water.util.FileUtils;
-import water.util.Log;
 
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
@@ -28,36 +27,25 @@
       }
       Frame dataset = DKV.getGet(f_name);
       Frame.CSVStreamParams parms = new Frame.CSVStreamParams()
-          .setHexString(Boolean.parseBoolean(hex_string))
-          .setEscapeQuotes(Boolean.parseBoolean(escape_quotes_string));
-      InputStream is = dataset.toCSV(parms);
-      response.setContentType("application/octet-stream");
-      // Clean up the file name
-      int x = f_name.length() - 1;
-      boolean dot = false;
-      for (; x >= 0; x--)
-        if (!Character.isLetterOrDigit(f_name.charAt(x)) && f_name.charAt(x) != '_')
-          if (f_name.charAt(x) == '.' && !dot) dot = true;
-          else break;
-      String suggested_fname = f_name.substring(x + 1).replace(".hex", ".csv");
-      if (!suggested_fname.endsWith(".csv"))
-        suggested_fname = suggested_fname + ".csv";
-      f_name = suggested_fname;
-      response.addHeader("Content-Disposition", "attachment; filename=" + f_name);
-      ServletUtils.setResponseStatus(response, HttpServletResponse.SC_OK);
-      OutputStream os = null;
-      try {
-        os = response.getOutputStream();
+              .setHexString(Boolean.parseBoolean(hex_string))
+              .setEscapeQuotes(Boolean.parseBoolean(escape_quotes_string));
+
+      try (final InputStream is = dataset.toCSV(parms); final OutputStream os = response.getOutputStream()) {
+        response.setContentType("application/octet-stream");
+        // Clean up the file name
+        int x = f_name.length() - 1;
+        boolean dot = false;
+        for (; x >= 0; x--)
+          if (!Character.isLetterOrDigit(f_name.charAt(x)) && f_name.charAt(x) != '_')
+            if (f_name.charAt(x) == '.' && !dot) dot = true;
+            else break;
+        String suggested_fname = f_name.substring(x + 1).replace(".hex", ".csv");
+        if (!suggested_fname.endsWith(".csv"))
+          suggested_fname = suggested_fname + ".csv";
+        f_name = suggested_fname;
+        response.addHeader("Content-Disposition", "attachment; filename=" + f_name);
+        ServletUtils.setResponseStatus(response, HttpServletResponse.SC_OK);
         FileUtils.copyStream(is, os, 2048);
-      } finally {
-        if (os != null) {
-          try {
-            os.close();
-          }
-          catch (Exception e) {
-            Log.err(e);
-          }
-        }
       }
     } catch (Exception e) {
       ServletUtils.sendErrorResponse(response, e, uri);
```

Comment (Pscheidl) [reply h2o-core/src/main/java/water/api/DatasetServlet.java]: And when we're at it, maybe using `IOUtils.copy(is, os);` would serve us better than our own implementation with custom fixed buffer size (2k bytes might be too small in many cases ? ).

Just noticed that. It's a minor thing. We don't need to do it in this PR.

Comment (honzasterba) [reply h2o-core/src/main/java/water/api/DatasetServlet.java]: sounds good, but I would do it in separate PR

Comment (Pscheidl) [reply h2o-core/src/main/java/water/api/DatasetServlet.java]: Roger that sir ! No problem here.

**Comment by honzasterba** [comment]
@mattdowle @ledell please review, I have decided to use `.datatable.aware` since the only place we should get in touch with data.frames is as.h2o, any other concerns?

File: h2o-r/tests/testdir_jira/runit_pubdev_2844.R [hunk 2]
```diff
@@ -5,7 +5,9 @@ test.pubdev_2844 <- function() {
   
   df1 <- iris
   h2o.no_progress()
-  
+  # enable using of data.table for as.data.frame
+  options(h2o.fread=TRUE)
```
Comment (michalkurka) [inline h2o-r/tests/testdir_jira/runit_pubdev_2844.R]: does this also enable it for the rest of the test suite? or is it just this test?

Comment (honzasterba) [reply h2o-r/tests/testdir_jira/runit_pubdev_2844.R]: I think it probably does, will fix

**Comment by michalkurka** [comment]
@honzasterba looks good to me, but please change the target branch to master and lets target 3.30 for this feature

File: h2o-r/h2o-package/R/frame.R [hunk 3]
```diff
@@ -3530,6 +3530,8 @@ as.data.frame.H2OFrame <- function(x, ...) {
   # Versions of R prior to 3.1 should not use hex string.
   # Versions of R including 3.1 and later should use hex string.
   useHexString <- getRversion() >= "3.1"
+  # We cannot use data.table by default since its handling of escaping inside quoted csv values is not very good
+  # in some edge cases its simply impossible to load data in correct format without additional post processing
```
Comment (mattdowle) [inline h2o-r/h2o-package/R/frame.R]: What specifically do you mean by "not very good"? This is unhelpful to write in this way. Tell me what's wrong and I'll fix it. Probably already known somewhere and I'll raise the priority.  We're on the same team here.

Comment (honzasterba) [reply h2o-r/h2o-package/R/frame.R]: Sorry for being so ambiguous. On second thought I should have been more to the point and link to existing issue. 
This is the issue I was reffering to:
This is the issue: https://github.com/Rdatatable/data.table/issues/1109
Here is the test that I was not able to make pass with data.table even with changes on the back-end: https://github.com/h2oai/h2o-3/pull/4265/files#diff-c87a495277fa01b186f5300c48973bf6R33

**Pull request merged**